aws_spot_fleet_request: unexpected EOF (panic: interface conversion: interface is nil, not string)

### Terraform Version

`Terraform v0.7.3`
### Affected Resource(s)
- aws_spot_fleet_request
### Terraform Configuration Files

[Encrypted with Hashicorp GPG](https://www.dropbox.com/s/252x92r5pfanf6d/spark.tf.binary_hashicorp_gpg?dl=0)
### Debug Output

Just before it dies [it describes image to get the root and ephemeral block device configuration](https://gist.githubusercontent.com/liquid-sky/a042b8910b7a5abf0647849b2c894440/raw/e0f9df9322a0163b5e33905c4f508024fdc07f2f/debug.log)

[This is the last segment that gets successfully executed.](https://github.com/hashicorp/terraform/blob/v0.7.3/builtin/providers/aws/resource_aws_spot_fleet_request.go#L108)
### Panic Output

[Encrypted with Hashicorp GPG](https://www.dropbox.com/s/gfutdb6hxt3o8c5/crash.log.binary_hashicorp_gpg?dl=0)
### Expected Behavior

When creating Spot Fleet, Terraform crashes after the fleet has been created and before it saves the state. 

Terraform should launch Spot Fleet and save the state of configuration. 
### Actual Behavior

Terraform exits abruptly and abandons the state, however Spot Fleet gets created and TF quits while trying to describe configuration.  
### Steps to Reproduce
1. `terraform apply`

+1

Terraform v0.7.4
Terraform plan => ok
Terraform apply => panic: interface conversion: interface is nil, not string:
```
Error applying plan:

1 error(s) occurred:

* aws_spot_fleet_request.created_spot_fleet: unexpected EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
panic: interface conversion: interface is nil, not string
2016/09/23 11:51:25 [DEBUG] plugin: terraform:
2016/09/23 11:51:25 [DEBUG] plugin: terraform: goroutine 134 [running]:
2016/09/23 11:51:25 [DEBUG] plugin: terraform: panic(0x2523320, 0xc420419500)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/go/src/runtime/panic.go:500 +0x1a1
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.hashEphemeralBlockDevice(0x2474aa0, 0xc4207f1e30, 0xc4202096b8)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:955 +0x213
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Set).hash(0xc4202096a0, 0x2474aa0, 0xc4207f1e30, 0x0, 0x5b0000c42005c400)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/set.go:180 +0x3d
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Set).add(0xc4202096a0, 0x2474aa0, 0xc4207f1e30, 0x412cc00, 0xc4205cb160, 0xc4207a91a8)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/set.go:167 +0x81
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Set).Add(0xc4202096a0, 0x2474aa0, 0xc4207f1e30)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/set.go:69 +0x44
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.ebsBlockDevicesToSet(0xc4205d3770, 0x2, 0x2, 0xc4201323e8, 0xc420209660)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:848 +0x649
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.launchSpecToMap(0xc42032b5c0, 0xc4201323e8, 0xc42002e0f0)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:753 +0x155
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.launchSpecsToSet(0xc42002e400, 0x1, 0x1, 0xc42002e0f0, 0xc4205ba100)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:739 +0x158
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsSpotFleetRequestRead(0xc42037dd40, 0x21fcee0, 0xc420314180, 0x0, 0x0)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:727 +0x3cc
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/builtin/providers/aws.resourceAwsSpotFleetRequestCreate(0xc42037dd40, 0x21fcee0, 0xc420314180, 0x2528640, 0x15d81)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/builtin/providers/aws/resource_aws_spot_fleet_request.go:617 +0xc43
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc42014f500, 0xc4204b6640, 0xc4201465a0, 0x21fcee0, 0xc420314180, 0x1, 0xc4203c3230, 0x0)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/resource.go:162 +0x30e
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc420684960, 0xc420127f80, 0xc4204b6640, 0xc4201465a0, 0x0, 0x18, 0x18)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/helper/schema/provider.go:178 +0x9b
2016/09/23 11:51:25 [DEBUG] plugin: terraform: github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc42069a2c0, 0xc42068bdc0, 0xc420134880, 0x0, 0x0)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/gopath/src/github.com/hashicorp/terraform/plugin/resource_provider.go:460 +0x57
2016/09/23 11:51:25 [DEBUG] plugin: terraform: reflect.Value.call(0xc42049e3c0, 0xc4201b8000, 0x13, 0x2a63bf3, 0x4, 0xc4207a9ed0, 0x3, 0x3, 0x22d7320, 0xc42037daa0, ...)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/go/src/reflect/value.go:434 +0x5c8
2016/09/23 11:51:25 [DEBUG] plugin: terraform: reflect.Value.Call(0xc42049e3c0, 0xc4201b8000, 0x13, 0xc4207a9ed0, 0x3, 0x3, 0xc4207fe014, 0x100000000, 0xc42035d701)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/go/src/reflect/value.go:302 +0xa4
2016/09/23 11:51:25 [DEBUG] plugin: terraform: net/rpc.(*service).call(0xc420664200, 0xc4206641c0, 0xc4206a0240, 0xc420352300, 0xc4206ac9a0, 0x21fec20, 0xc42068bdc0, 0x16, 0x21fec60, 0xc420134880, ...)
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/go/src/net/rpc/server.go:383 +0x148
2016/09/23 11:51:25 [DEBUG] plugin: terraform: created by net/rpc.(*Server).ServeCodec
2016/09/23 11:51:25 [DEBUG] plugin: terraform: 	/opt/go/src/net/rpc/server.go:477 +0x421
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalWriteState
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalIf
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalWriteState
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalWriteDiff
2016/09/23 11:51:25 [DEBUG] root: eval: *terraform.EvalApplyPost
2016/09/23 11:51:25 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_spot_fleet_request.created_spot_fleet: unexpected EOF
2016/09/23 11:51:25 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_spot_fleet_request.created_spot_fleet: unexpected EOF
2016/09/23 11:51:25 [ERROR] root: eval: *terraform.EvalOpFilter, err: 1 error(s) occurred:

* aws_spot_fleet_request.created_spot_fleet: unexpected EOF
2016/09/23 11:51:25 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_spot_fleet_request.created_spot_fleet: unexpected EOF
2016/09/23 11:51:25 [TRACE] [walkApply] Exiting eval tree: aws_spot_fleet_request.created_spot_fleet
2016/09/23 11:51:25 [DEBUG] vertex provider.aws (close), got dep: aws_spot_fleet_request.created_spot_fleet
2016/09/23 11:51:25 [DEBUG] vertex root, got dep: provider.aws (close)
2016/09/23 11:51:25 [DEBUG] vertex root, got dep: output.templates
2016/09/23 11:51:25 [DEBUG] plugin: /Users/andrey/work/bin/terraform-0.7.4/terraform: plugin process exited
2016/09/23 11:51:25 [TRACE] Preserving existing state lineage "914c60d1-6556-409e-a495-7c6ead77edbf"
2016/09/23 11:51:25 [TRACE] Preserving existing state lineage "914c60d1-6556-409e-a495-7c6ead77edbf"
2016/09/23 11:51:25 [DEBUG] plugin: waiting for all plugin processes to complete...
2016/09/23 11:51:25 [DEBUG] plugin: /Users/andrey/work/bin/terraform-0.7.4/terraform: plugin process exited
2016/09/23 11:51:25 [DEBUG] plugin: /Users/andrey/work/bin/terraform-0.7.4/terraform: plugin process exited
```

--- a/builtin/providers/aws/resource_aws_spot_fleet_request.go
+++ b/builtin/providers/aws/resource_aws_spot_fleet_request.go
@@ -805,7 +805,7 @@ func launchSpecToMap(
 }
 
 func ebsBlockDevicesToSet(bdm []*ec2.BlockDeviceMapping, rootDevName *string) *schema.Set {
-	set := &schema.Set{F: hashEphemeralBlockDevice}
+	set := &schema.Set{F: hashEbsBlockDevice}
 
 	for _, val := range bdm {
 		if val.Ebs != nil {
@@ -1009,7 +1009,11 @@ func hashLaunchSpecification(v interface{}) int {
 func hashEbsBlockDevice(v interface{}) int {
 	var buf bytes.Buffer
 	m := v.(map[string]interface{})
-	buf.WriteString(fmt.Sprintf("%s-", m["device_name"].(string)))
-	buf.WriteString(fmt.Sprintf("%s-", m["snapshot_id"].(string)))
+	if name, ok := m["device_name"]; ok {
+		buf.WriteString(fmt.Sprintf("%s-", name.(string)))
+	}
+	if id, ok := m["snapshot_id"]; ok {
+		buf.WriteString(fmt.Sprintf("%s-", id.(string)))
+	}
 	return hashcode.String(buf.String())
 }

--- a/builtin/providers/aws/resource_aws_spot_fleet_request_test.go
+++ b/builtin/providers/aws/resource_aws_spot_fleet_request_test.go
@@ -293,6 +293,28 @@ func TestAccAWSSpotFleetRequest_withWeightedCapacity(t *testing.T) {
 	})
 }
 
+func TestAccAWSSpotFleetRequest_withEBSDisk(t *testing.T) {
+	var config ec2.SpotFleetRequestConfig
+	rName := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSSpotFleetRequestDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSSpotFleetRequestEBSConfig(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSSpotFleetRequestExists(
+						"aws_spot_fleet_request.foo", &config),
+					testAccCheckAWSSpotFleetRequest_EBSAttributes(
+						&config),
+				),
+			},
+		},
+	})
+}
+
 func TestAccAWSSpotFleetRequest_CannotUseEmptyKeyName(t *testing.T) {
 	_, errors := validateSpotFleetRequestKeyName("", "key_name")
 	if len(errors) == 0 {
@@ -370,6 +392,31 @@ func testAccCheckAWSSpotFleetRequest_LaunchSpecAttributes(
 
 		if *spec.UserData != base64.StdEncoding.EncodeToString([]byte("hello-world")) {
 			return fmt.Errorf("Unexpected launch specification user data: %s", *spec.UserData)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckAWSSpotFleetRequest_EBSAttributes(
+	sfr *ec2.SpotFleetRequestConfig) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if len(sfr.SpotFleetRequestConfig.LaunchSpecifications) == 0 {
+			return fmt.Errorf("Missing launch specification")
+		}
+
+		spec := *sfr.SpotFleetRequestConfig.LaunchSpecifications[0]
+
+		ebs := spec.BlockDeviceMappings
+		if len(ebs) < 2 {
+			return fmt.Errorf("Expected %d block device mappings, got %d", 2, len(ebs))
+		}
+
+		if *ebs[0].DeviceName != "/dev/xvda" {
+			return fmt.Errorf("Expected device 0's name to be %s, got %s", "/dev/xvda", *ebs[0].DeviceName)
+		}
+		if *ebs[1].DeviceName != "/dev/xvdcz" {
+			return fmt.Errorf("Expected device 1's name to be %s, got %s", "/dev/xvdcz", *ebs[1].DeviceName)
 		}
 
 		return nil
@@ -939,4 +986,61 @@ resource "aws_spot_fleet_request" "foo" {
     depends_on = ["aws_iam_policy_attachment.test-attach"]
 }
 `, rName, rName)
+}
+
+func testAccAWSSpotFleetRequestEBSConfig(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_iam_policy_attachment" "test-attach" {
+    name = "test-attachment"
+    roles = ["${aws_iam_role.test-role.name}"]
+    policy_arn = "arn:aws:iam::aws:policy/service-role/AmazonEC2SpotFleetRole"
+}
+
+resource "aws_iam_role" "test-role" {
+    name = "test-role-%s"
+    assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Principal": {
+        "Service": [
+          "spotfleet.amazonaws.com",
+          "ec2.amazonaws.com"
+        ]
+      },
+      "Action": "sts:AssumeRole"
+    }
+  ]
+}
+EOF
+}
+
+resource "aws_spot_fleet_request" "foo" {
+    iam_fleet_role = "${aws_iam_role.test-role.arn}"
+    spot_price = "0.005"
+    target_capacity = 1
+    valid_until = "2019-11-04T20:44:20Z"
+    terminate_instances_with_expiration = true
+    launch_specification {
+        instance_type = "m1.small"
+        ami = "ami-d06a90b0"
+
+	ebs_block_device {
+            device_name = "/dev/xvda"
+	    volume_type = "gp2"
+	    volume_size = "8"
+        }
+	
+	ebs_block_device {
+            device_name = "/dev/xvdcz"
+	    volume_type = "gp2"
+	    volume_size = "100"
+        }
+    }
+    depends_on = ["aws_iam_policy_attachment.test-attach"]
+}
+`, rName)
 }

